File: python-sdks/instrumentations/respan-instrumentation-aws-bedrock/src/respan_instrumentation_aws_bedrock/_otel_emitter.py
"""Emit AWS Bedrock Runtime calls as OTEL ReadableSpan objects."""

from __future__ import annotations

import logging
import time
from collections.abc import Mapping
from typing import Any

from opentelemetry import context as context_api
from opentelemetry import trace
from opentelemetry.semconv._incubating.attributes import (
    gen_ai_attributes as GenAIAttributes,
)
from opentelemetry.semconv_ai import LLMRequestTypeValues
from opentelemetry.semconv_ai import SpanAttributes as TLSpanAttributes

from respan_instrumentation_aws_bedrock._constants import (
    AWS_BEDROCK_CHAT_SPAN_NAME,
    AWS_BEDROCK_SYSTEM_NAME,
)
from respan_instrumentation_aws_bedrock._translator import (
    BedrockResponse,
    parse_bedrock_request,
    parse_bedrock_response,
    parse_bedrock_stream_response,
    safe_json,
    to_json_attr,
)
from respan_sdk.constants.llm_logging import LOG_TYPE_CHAT
from respan_sdk.constants.span_attributes import RESPAN_LOG_TYPE
from respan_sdk.utils.data_processing.id_processing import (
    format_span_id,
    format_trace_id,
)
from respan_tracing.utils.span_factory import build_readable_span, inject_span

logger = logging.getLogger(__name__)


def _current_trace_parent_ids() -> tuple[str | None, str | None]:
    try:
        current_span = trace.get_current_span()
        span_context = current_span.get_span_context()
    except Exception:
        return None, None

    trace_id = getattr(span_context, "trace_id", 0) or 0
    span_id = getattr(span_context, "span_id", 0) or 0
    if trace_id == 0 or span_id == 0:
        return None, None
    return format_trace_id(trace_id=trace_id), format_span_id(span_id=span_id)


def _base_attrs() -> dict[str, Any]:
    attrs = {
        TLSpanAttributes.LLM_SYSTEM: AWS_BEDROCK_SYSTEM_NAME,
        TLSpanAttributes.LLM_REQUEST_TYPE: LLMRequestTypeValues.CHAT.value,
        TLSpanAttributes.TRACELOOP_ENTITY_NAME: AWS_BEDROCK_CHAT_SPAN_NAME,
        TLSpanAttributes.TRACELOOP_ENTITY_PATH: AWS_BEDROCK_CHAT_SPAN_NAME,
        RESPAN_LOG_TYPE: LOG_TYPE_CHAT,
    }
    workflow_name = context_api.get_value(TLSpanAttributes.TRACELOOP_ENTITY_NAME)
    if workflow_name:
        attrs[TLSpanAttributes.TRACELOOP_WORKFLOW_NAME] = workflow_name
    return attrs


def _set_prompt_attrs(attrs: dict[str, Any], messages: list[dict[str, Any]]) -> None:
    for index, message in enumerate(messages):
        role = message.get("role")
        content = message.get("content")
        if role is not None:
            attrs[f"{TLSpanAttributes.LLM_PROMPTS}.{index}.role"] = str(role)
        if content is not None:
            attrs[f"{TLSpanAttributes.LLM_PROMPTS}.{index}.content"] = to_json_attr(
                content
            )


def _set_usage_attrs(attrs: dict[str, Any], usage: Mapping[str, int]) -> None:
    input_tokens = usage.get("input_tokens")
    output_tokens = usage.get("output_tokens")
    total_tokens = usage.get("total_tokens")

    if input_tokens is not None:
        attrs[GenAIAttributes.GEN_AI_USAGE_INPUT_TOKENS] = input_tokens
        attrs[TLSpanAttributes.LLM_USAGE_PROMPT_TOKENS] = input_tokens
    if output_tokens is not None:
        attrs[GenAIAttributes.GEN_AI_USAGE_OUTPUT_TOKENS] = output_tokens
        attrs[TLSpanAttributes.LLM_USAGE_COMPLETION_TOKENS] = output_tokens
    if total_tokens is not None:
        attrs[TLSpanAttributes.LLM_USAGE_TOTAL_TOKENS] = total_tokens


def _set_response_attrs(attrs: dict[str, Any], response: BedrockResponse) -> None:
    attrs[f"{TLSpanAttributes.LLM_COMPLETIONS}.0.role"] = response.role
    attrs[f"{TLSpanAttributes.LLM_COMPLETIONS}.0.content"] = response.content
    attrs[TLSpanAttributes.TRACELOOP_ENTITY_OUTPUT] = response.content
    if response.tool_calls:
        attrs[f"{TLSpanAttributes.LLM_COMPLETIONS}.0.tool_calls"] = safe_json(
            response.tool_calls
        )
    _set_usage_attrs(attrs=attrs, usage=response.usage)


def build_bedrock_attrs(
    *,
    operation_name: str,
    api_params: Mapping[str, Any] | None,
    response_payload: Any = None,
    stream_events: list[Any] | None = None,
) -> dict[str, Any]:
    attrs = _base_attrs()
    request = parse_bedrock_request(
        operation_name=operation_name,
        api_params=api_params,
    )

    if request.model_id:
        attrs[TLSpanAttributes.LLM_REQUEST_MODEL] = request.model_id

    if request.messages:
        attrs[TLSpanAttributes.TRACELOOP_ENTITY_INPUT] = safe_json(request.messages)
        _set_prompt_attrs(attrs=attrs, messages=request.messages)
    elif request.raw_payload is not None:
        attrs[TLSpanAttributes.TRACELOOP_ENTITY_INPUT] = safe_json(request.raw_payload)

    if request.tools:
        attrs[TLSpanAttributes.LLM_REQUEST_FUNCTIONS] = safe_json(request.tools)

    response: BedrockResponse | None = None
    if stream_events is not None:
        response = parse_bedrock_stream_response(
            operation_name=operation_name,
            events=stream_events,
        )
    elif response_payload is not None:
        response = parse_bedrock_response(
            operation_name=operation_name,
            response_payload=response_payload,
        )

    if response is not None:
        _set_response_attrs(attrs=attrs, response=response)

    return attrs


def emit_bedrock_span(
    *,
    operation_name: str,
    api_params: Mapping[str, Any] | None,
    start_ns: int,
    response_payload: Any = None,
    stream_events: list[Any] | None = None,
    error_message: str | None = None,
    status_code: int = 200,
) -> None:
    """Build a ReadableSpan for an AWS Bedrock Runtime call and inject it."""
    try:
        attrs = build_bedrock_attrs(
            operation_name=operation_name,
            api_params=api_params,
            response_payload=response_payload,
            stream_events=stream_events,
        )
        if error_message:
            attrs["error.message"] = error_message
            status_code = status_code if status_code >= 400 else 500

        trace_id, parent_id = _current_trace_parent_ids()
        span = build_readable_span(
            name=AWS_BEDROCK_CHAT_SPAN_NAME,
            trace_id=trace_id,
            parent_id=parent_id,
            start_time_ns=start_ns,
            end_time_ns=time.time_ns(),
            attributes=attrs,
            error_message=error_message,
            status_code=status_code,
        )
        inject_span(span=span)
    except Exception:
        logger.debug("Failed to emit AWS Bedrock span", exc_info=True)

File: javascript-sdks/instrumentations/respan-instrumentation-aws-bedrock/src/_otel_emitter.ts
import { context, trace, TraceFlags } from "@opentelemetry/api";
import { hrTime } from "@opentelemetry/core";
import type { ReadableSpan } from "@opentelemetry/sdk-trace-base";
import {
  ATTR_ERROR_MESSAGE,
  ATTR_GEN_AI_REQUEST_MODEL,
  ATTR_GEN_AI_SYSTEM,
  ATTR_GEN_AI_USAGE_INPUT_TOKENS,
  ATTR_GEN_AI_USAGE_OUTPUT_TOKENS,
  ATTR_GEN_AI_USAGE_PROMPT_TOKENS,
  ATTR_GEN_AI_USAGE_COMPLETION_TOKENS,
} from "@opentelemetry/semantic-conventions/incubating";
import { RespanLogType, RespanSpanAttributes } from "@respan/respan-sdk";
import { buildReadableSpan, injectSpan } from "@respan/tracing";
import {
  LLMRequestTypeValues,
  SpanAttributes,
} from "@traceloop/ai-semantic-conventions";
import {
  AWS_BEDROCK_CHAT_SPAN_NAME,
  AWS_BEDROCK_INSTRUMENTATION_PACKAGE,
  AWS_BEDROCK_SYSTEM_NAME,
  PACKAGE_VERSION,
  RESPAN_LOG_METHOD_TS_TRACING,
  STATUS_CODE_ATTR,
} from "./_constants.js";
import {
  type BedrockResponse,
  captureInvokeResponsePayload,
  parseBedrockRequest,
  parseBedrockResponse,
  parseBedrockStreamResponse,
  safeJson,
  toJsonAttr,
} from "./_translator.js";

type HrTime = [number, number];

export interface EmitBedrockSpanOptions {
  operationName: string;
  apiParams?: Record<string, unknown>;
  startTimeHr: HrTime;
  responsePayload?: unknown;
  streamEvents?: unknown[];
  errorMessage?: string;
  statusCode?: number;
}

function baseAttrs(): Record<string, unknown> {
  const attrs: Record<string, unknown> = {
    [ATTR_GEN_AI_SYSTEM]: AWS_BEDROCK_SYSTEM_NAME,
    [SpanAttributes.LLM_REQUEST_TYPE]: LLMRequestTypeValues.CHAT,
    [SpanAttributes.TRACELOOP_ENTITY_NAME]: AWS_BEDROCK_CHAT_SPAN_NAME,
    [SpanAttributes.TRACELOOP_ENTITY_PATH]: AWS_BEDROCK_CHAT_SPAN_NAME,
    [RespanSpanAttributes.RESPAN_LOG_METHOD]: RESPAN_LOG_METHOD_TS_TRACING,
    [RespanSpanAttributes.RESPAN_LOG_TYPE]: RespanLogType.CHAT,
  };

  const workflowName = (context.active() as any).getValue(SpanAttributes.TRACELOOP_ENTITY_NAME);
  if (typeof workflowName === "string" && workflowName) {
    attrs[SpanAttributes.TRACELOOP_WORKFLOW_NAME] = workflowName;
  }
  return attrs;
}

function setPromptAttrs(
  attrs: Record<string, unknown>,
  messages: Record<string, unknown>[],
): void {
  messages.forEach((message, index) => {
    const prefix = `${SpanAttributes.LLM_PROMPTS}.${index}`;
    const role = message.role;
    const content = message.content;
    if (role !== undefined) {
      attrs[`${prefix}.role`] = String(role);
    }
    if (content !== undefined) {
      attrs[`${prefix}.content`] = toJsonAttr(content);
    }
    if (message.tool_calls !== undefined) {
      attrs[`${prefix}.tool_calls`] = safeJson(message.tool_calls);
    }
    if (message.tool_call_id !== undefined) {
      attrs[`${prefix}.tool_call_id`] = String(message.tool_call_id);
    }
  });
}

function setUsageAttrs(
  attrs: Record<string, unknown>,
  usage: Record<string, number>,
): void {
  const inputTokens = usage.input_tokens;
  const outputTokens = usage.output_tokens;
  const totalTokens = usage.total_tokens;

  if (inputTokens !== undefined) {
    attrs[ATTR_GEN_AI_USAGE_INPUT_TOKENS] = inputTokens;
    attrs[ATTR_GEN_AI_USAGE_PROMPT_TOKENS] = inputTokens;
  }
  if (outputTokens !== undefined) {
    attrs[ATTR_GEN_AI_USAGE_OUTPUT_TOKENS] = outputTokens;
    attrs[ATTR_GEN_AI_USAGE_COMPLETION_TOKENS] = outputTokens;
  }
  if (totalTokens !== undefined) {
    attrs[SpanAttributes.LLM_USAGE_TOTAL_TOKENS] = totalTokens;
  }
}

function setResponseAttrs(
  attrs: Record<string, unknown>,
  response: BedrockResponse,
): void {
  attrs[`${SpanAttributes.LLM_COMPLETIONS}.0.role`] = response.role;
  attrs[`${SpanAttributes.LLM_COMPLETIONS}.0.content`] = response.content;
  attrs[SpanAttributes.TRACELOOP_ENTITY_OUTPUT] = response.content;
  if (response.toolCalls.length > 0) {
    attrs[`${SpanAttributes.LLM_COMPLETIONS}.0.tool_calls`] = safeJson(response.toolCalls);
  }
  setUsageAttrs(attrs, response.usage);
}

export function buildBedrockAttrs(options: {
  operationName: string;
  apiParams?: Record<string, unknown>;
  responsePayload?: unknown;
  streamEvents?: unknown[];
}): Record<string, unknown> {
  const attrs = baseAttrs();
  const request = parseBedrockRequest({
    operationName: options.operationName,
    apiParams: options.apiParams,
  });

  if (request.modelId) {
    attrs[ATTR_GEN_AI_REQUEST_MODEL] = request.modelId;
  }

  if (request.messages.length > 0) {
    attrs[SpanAttributes.TRACELOOP_ENTITY_INPUT] = safeJson(request.messages);
    setPromptAttrs(attrs, request.messages);
  } else if (request.rawPayload !== undefined) {
    attrs[SpanAttributes.TRACELOOP_ENTITY_INPUT] = safeJson(request.rawPayload);
  }

  if (request.tools.length > 0) {
    attrs[SpanAttributes.LLM_REQUEST_FUNCTIONS] = safeJson(request.tools);
  }

  let response: BedrockResponse | undefined;
  if (options.streamEvents !== undefined) {
    response = parseBedrockStreamResponse({
      operationName: options.operationName,
      events: options.streamEvents,
    });
  } else if (options.responsePayload !== undefined) {
    response = parseBedrockResponse({
      operationName: options.operationName,
      responsePayload: options.responsePayload,
    });
  }

  if (response) {
    setResponseAttrs(attrs, response);
  }
  return attrs;
}

function activeSpanContext(): ReturnType<ReadableSpan["spanContext"]> | undefined {
  return trace.getSpan(context.active())?.spanContext();
}

function buildInstrumentedReadableSpan(
  options: Parameters<typeof buildReadableSpan>[0],
): ReadableSpan {
  const activeContext = activeSpanContext();
  const span = buildReadableSpan(options) as ReadableSpan & {
    instrumentationScope?: {
      name: string;
      version?: string;
    };
    spanContext: () => ReturnType<ReadableSpan["spanContext"]>;
  };

  const originalSpanContext = span.spanContext.bind(span);
  span.spanContext = () => ({
    ...originalSpanContext(),
    traceFlags: activeContext?.traceFlags ?? TraceFlags.SAMPLED,
  });
  span.instrumentationScope = {
    name: AWS_BEDROCK_INSTRUMENTATION_PACKAGE,
    version: PACKAGE_VERSION,
  };
  return span;
}

export function emitBedrockSpan(options: EmitBedrockSpanOptions): void {
  try {
    const attrs = buildBedrockAttrs({
      operationName: options.operationName,
      apiParams: options.apiParams,
      responsePayload: options.responsePayload,
      streamEvents: options.streamEvents,
    });
    let statusCode = options.statusCode ?? 200;
    if (options.errorMessage) {
      attrs[ATTR_ERROR_MESSAGE] = options.errorMessage;
      statusCode = statusCode >= 400 ? statusCode : 500;
    }
    attrs[STATUS_CODE_ATTR] = statusCode;

    const activeContext = activeSpanContext();
    const span = buildInstrumentedReadableSpan({
      name: AWS_BEDROCK_CHAT_SPAN_NAME,
      traceId: activeContext?.traceId,
      parentId: activeContext?.spanId,
      startTimeHr: options.startTimeHr,
      endTimeHr: hrTime(),
      attributes: attrs,
      statusCode,
      errorMessage: options.errorMessage,
    });
    injectSpan(span);
  } catch {
    // Instrumentation must not break application code.
  }
}

export function responsePayloadForInvoke(response: unknown): unknown {
  return captureInvokeResponsePayload(response);
}
